numpy/core/src/multiarray/arrayobject.c

```diff
@@ -766,7 +766,15 @@ _void_compare(PyArrayObject *self, PyArrayObject *other, int cmp_op)
                         memcpy(dimensions, PyArray_DIMS((PyArrayObject *)temp),
                                sizeof(npy_intp)*result_ndim);
                     }
-                    dimensions[result_ndim] = -1;
+
+                    /*
+                    * Compute the new dimension size manually 
+                    * (instead of reshaping with -1) to avoid division by zero.
```

Comment:
Might just say that `-1` will not work when the array is empty (gh-18519).  Division by zero seems is the cause for the failure, but it not the problematic condition at this place (the empty array is).

Otherwise, looks great though!  A small update would be great, or we may just put it in (or do a small update).

Reply:
Updated the comment, thanks!